defaults.go

```diff
@@ -272,18 +272,24 @@ func IsMsg(buf []byte) error {
 
 // IsFqdn checks if a domain name is fully qualified.
 func IsFqdn(s string) bool {
-	s2 := strings.TrimSuffix(s, ".")
-	if s == s2 {
+	// Check for (and remove) a trailing dot, returning if there isn't one.
+	if s == "" || s[len(s)-1] != '.' {
 		return false
 	}
+	s = s[:len(s)-1]
 
-	i := strings.LastIndexFunc(s2, func(r rune) bool {
+	// If we don't have an escape sequence before the final dot, we know it's
+	// fully qualified and can return here.
+	if s == "" || s[len(s)-1] != '\\' {
```

Comment:
It might be faster still if a new slice is only constructed in the slow path.
```go
// IsFqdn checks if a domain name is fully qualified
// (specifically, if it ends with an unescaped dot).
func IsFqdn(s string) bool {
	// If there's no trailing dot, the input is definitely not fully qualified.
	if s == "" || s[len(s)-1] != '.' {
		return false
	}

	// If there's no backslash before the final dot,
	// it's definitely not escaped so assume the input is fully qualified.
	if len(s) == 1 || s[len(s)-2] != '\\' {
		return true
	}

	// Otherwise the dot is escaped iff the count of backslashes before the
	// one we already found is even (including if there are none).
	prefix := s[:len(s)-2]
	i := strings.LastIndexFunc(prefix, func(r rune) bool {
		return r != '\\'
	})
	return (len(prefix)-i)%2 == 0
}
```

Reply:
It actually ends up slower on my system:
```
name                 old time/op    new time/op    delta
IsFQDN/no_dot-12       1.48ns ± 3%    1.54ns ± 1%  +3.90%  (p=0.000 n=10+9)
IsFQDN/unescaped-12    1.57ns ± 1%    1.66ns ± 3%  +5.37%  (p=0.000 n=8+9)
IsFQDN/escaped-12      23.8ns ± 2%    24.4ns ± 2%  +2.58%  (p=0.000 n=10+10)
FQDN/is_fqdn-12        1.80ns ± 2%    1.87ns ± 3%  +4.09%  (p=0.000 n=10+9)
FQDN/not_fqdn-12       33.4ns ±12%    32.5ns ±13%    ~     (p=0.315 n=10+10)
```

Reply:
Ok. My other thought is special-casing `if len(s) <= 1 || s[len(s)-1] != '.' { return s == "." }` to avoid a comparison in the nontrivial-dot-terminated-string case, but it's probably best to avoid drowning in micro-optimization possibilities.